lib/services/cloud-codesign-service.ts

```diff
@@ -0,0 +1,133 @@
+import { CloudService } from "./cloud-service";
+
+export class CloudCodesignService extends CloudService implements ICloudCodesignService {
+	private platform = this.$devicePlatformsConstants.iOS;
+
+	protected failedError: string;
+	protected failedToStartError: string;
+	protected operationInProgressStatus: string;
+
+	constructor($fs: IFileSystem,
+		$httpClient: Server.IHttpClient,
+		$logger: ILogger,
+		private $buildCloudService: IBuildCloudService,
+		private $errors: IErrors,
+		private $devicesService: Mobile.IDevicesService,
+		private $projectHelper: IProjectHelper,
+		private $devicePlatformsConstants: Mobile.IDevicePlatformsConstants) {
+		super($fs, $httpClient, $logger);
+		this.failedToStartError = "Failed to start cloud codesign generation.";
+		this.failedError = "Codesign generation failed.";
+		this.operationInProgressStatus = "Working";
+	}
+
+	public async generateCodesignFiles(codesignData: ICodesignData,
+		projectData: IProjectData): Promise<ICodesignResultData> {
+		codesignData.clean = codesignData.clean === undefined ? true : codesignData.clean;
+		const buildId = uuid.v4();
+
+		try {
+			const serverResult = await this.executeGeneration(codesignData, projectData, buildId);
+			return serverResult;
+		} catch (err) {
+			err.buildId = buildId;
+			throw err;
+		}
+	}
+
+	public getServerOperationOutputDirectory(options: ICloudServerOutputDirectoryOptions): string {
+		return path.join(options.projectDir, constants.CODESIGN_TEMP_DIR_NAME, options.platform.toLowerCase());
+	}
+
+	protected getServerResults(codesignResult: IServerResult): IServerItem[] {
+		const result = _.filter(codesignResult.buildItems, b => b.disposition === constants.DISPOSITIONS.CERTIFICATE
+			|| b.disposition === constants.DISPOSITIONS.PROVISION);
+
+		if (!result) {
+			this.$errors.failWithoutHelp(
+				`No item with disposition ${constants.DISPOSITIONS.CERTIFICATE} or ${constants.DISPOSITIONS.PROVISION} found in the server result items.`);
+		}
+
+		return result;
+	}
+
+	protected async getServerLogs(logsUrl: string, buildId: string): Promise<void> {
+		// no specific implementation needed.
+	}
+
+	public async executeGeneration(codesignData: ICodesignData,
+		projectData: IProjectData,
+		buildId: string): Promise<ICodesignResultData> {
+		const codesignInformationString = "Generation of iOS certificate and provision files";
+		this.$logger.info(`Starting ${codesignInformationString}.`);
+
+		const codesignRequest = await this.prepareCodesignRequest(buildId, codesignData, projectData);
+		const codesignResponse: IServerResponse = await this.$buildCloudService.generateCodesignFiles(codesignRequest);
+		this.$logger.trace(`Codesign response: ${JSON.stringify(codesignResponse)}`);
+
+		let codesignResult: IServerResult;
+		try {
+			await this.waitForServerOperationToFinish(buildId, codesignResponse);
+		} catch (ex) {
+			this.$logger.trace("Codesign generation failed with err: ", ex);
+		} finally {
```

Comment:
what's the purpose of this `finally` - we catch the error, so you can safely remove the finally block and directly set:
```TypeScript
const codesignResult = await this.getObjectFromS3File<IServerResult>(codesignResponse.resultUrl);
```

Reply:
The idea here is that unlike the build lambda, the codesign one is using only two files to share what happened. I.e. the resultsUrl consists of all input and result items if available. We catch error that is thrown in the try, but if there is some error in the build machine scripts it will be written in the results file. So we always want to read the file. Otherwise we do not see why the build machine script failed. For example providing wrong credentials will come as stderr and will be logged in the results file. [Here](https://github.com/NativeScript/nativescript-cloud/pull/52/files/44396757a1f7ad7e8ff3c65c9e7fa268c89d9e69#diff-5cdce4f47c4d43f2b52f9f9b4f256d1dR83) you can see that we need the codesignResult in any case.

Reply:
I understand this, but I still do not see why you need `finally`, you can just use:
```TypeScript
try {
} catch (err) 
...
}

const codesignResult = await this.getObjectFromS3File<IServerResult>(codesignResponse.resultUrl);
```